File: Olympiad/POI/official/2006/okr/prog/okrs1.pas
(*************************************************************************
 *                                                                       *
 *                   XIII Olimpiada Informatyczna                        *
 *                                                                       *
 *   Rozwi-zanie zadania: OKR (Okresy s³ów)                              *
 *   Plik:                okrs1.pas                                      *
 *   Autor:               Karol Cwalina                                  *
 *   Opis:                Rozwiazanie brutalnie poprawne, czas O(n^3)    *
 *                                                                       *
 *************************************************************************)

program okresy_slow;
const MaxN=1000000;
var i,j,k,n : LongInt;
    odp : Int64;
    ok : Boolean;
    tekst : array [1..MaxN] of Char;

begin
    Readln(n);
    for i := 1 to n do Read(tekst[i]);

    odp := 0;
    for i := 1 to n do
        for j := i downto (i+3) div 2 do begin
            ok := True;
            for k := 0 to i-j do
                if tekst[k+1] <> tekst[k+j] then begin
                    ok := False;
                    break;
                end;
            if ok then begin
                odp := odp+j-1;
                break;
            end;
        end;

    Writeln (odp);
end.

File: Olympiad/POI/official/2006/code/okr/prog/okrs1.cpp
/*************************************************************************
 *                                                                       *
 *                   XIII Olimpiada Informatyczna                        *
 *                                                                       *
 *   Rozwi-zanie zadania: OKR (Okresy s³ów)                              *
 *   Plik:                okrs0.cpp                                      *
 *   Autor:               Karol Cwalina                                  *
 *   Opis:                Rozwiazanie brutalnie poprawne, czas O(n^3)    *
 *                                                                       *
 *************************************************************************/

#include <cstdio>

using namespace std;

int main (void) {
    int n;
    long long odp=0;
    scanf("%d", &n);
    char *tekst = new char[n+1];
    scanf("%s", tekst);

    /* Dla kazdego prefiksu tekstu szukamy jego maksymalnego okresu */
    for (int i=0; i<n; ++i)
        /* j jest testowana dlugoscia okresu */
        for (int j=i; i <= 2*j; --j) {
            /* Sprawdzamy, czy to j faktycznie odpowiada dobremu okresowi */
            bool ok=true;
            for (int k=0; k <= i-j; ++k)
                if (tekst[k] != tekst[j+k]) {
                   ok=false;
                   break;
                }
            if (ok) {
               odp += j;
               break;
            }
        }

    printf("%lld\n", odp);

    return 0;
}
